🎨 Enhanced Visual Diff System for Card Description Changes (Resolves #7201)

LexioJ opened the pull request:
## 🎯 Overview

This PR introduces a comprehensive visual diff system that transforms how card description changes are displayed in the activity feed, making it easier for users to understand what exactly changed in their content.

**Resolves #7201**

## ✨ Features

### 🔍 Smart Change Detection
- **Intelligent Move Detection (🚚)**: Identifies when lines are relocated without modification
- **Word-Level Modifications (✏️)**: Highlights specific word changes within lines using combined `<del>`/`<ins>` HTML tags for cleaner output
- **Clean Additions/Deletions (✨/🗑️)**: Clear indicators for new or removed content
- **Checkbox State Tracking**: Special handling for task list changes with visual emoji transitions
- **Quote Line Handling**: Properly formats blockquotes with ❞ emoji (fixes `&gt;` display issues)
- **Callout Block Transitions**: Shows emoji transitions for callout type changes (ℹ️→❗, ✅→⚠️, etc.)
- **Code Block Indicators**: Displays code block markers with ‹› symbol

### 🎭 Professional Visual Indicators
- **🚚** Move operations: *"🚚5 Line content (from 2)"*
- **✏️** Modifications with combined tags: *"✏️3 <del>Old content</del><ins>New content</ins>"*
- **✨** Additions: *"✨2 New content"*
- **🗑️** Deletions: *"🗑️4 Removed content"*
- **Special Line Types**:
  - Checkboxes: *"☑️→🔲"* or *"🔲→☑️"*
  - Callouts: *"ℹ️→❗"* (info to error)
  - Quotes: *"→❞ Quoted text"*
  - Code blocks: *"→‹›"*

### 🧠 Intelligent Filtering & Optimization
- Filters out unchanged content to reduce noise
- Ignores empty line operations that don't add value
- **Combines consecutive `<ins>` and `<del>` tags** to reduce visual clutter
- **Merges whitespace between tags** for cleaner output (e.g., `<del>text 1 text 2</del>` instead of `<del>text 1</del> <del>text 2</del>`)
- Provides line number context for better orientation
- **Activity feed compatible**: Uses plain text formatting that displays correctly without HTML escaping issues

## 🛠 Technical Implementation

### Core Components
1. **`DiffService`**: Service implementing LCS-based diff algorithm with multi-pass detection
2. **Enhanced `DeckProvider`**: Updated activity provider with dependency injection for DiffService
3. **Smart Algorithm Flow**:
   - **Pass 1**: Content analysis and operation extraction
   - **Pass 2**: Move detection (exact matches at different positions)
   - **Pass 3**: Modification detection with special line handling:
     - Checkbox state changes
     - Callout block type changes
     - Quote line changes (prevents `&gt;` issues)
   - **Pass 4**: Word-level diff with combined HTML tags
   - **Pass 5**: Remaining operations (pure adds/deletes)

### Algorithm Highlights
- **Longest Common Subsequence (LCS)** for optimal diff detection
- **Similarity scoring system** for intelligent operation pairing
- **Multi-pass processing** ensures accurate classification
- **Tag combination logic** merges consecutive `<ins>`/`<del>` tags and whitespace-only keep operations
- **HTML-safe output** compatible with Nextcloud's activity feed constraints

### Activity Feed Compatibility Fixes
- Removed reliance on HTML rendering in activity feed
- Special handling for `>` character to prevent `&gt;` display issues
- Quote lines detected and formatted with emoji before HTML escaping
- All special patterns handled with plain text + emoji approach

## 📸 Visual Examples

### Word-Level Changes
**Before**: `John Doe changed the description`
**After**: `✏️2 Update <del>deadline</del><ins>timeline</ins>`

### Checkbox Changes  
**Before**: `John Doe changed the description`
**After**: `✏️1 🔲→☑️ Task completed`

### Callout Block Changes
**Before**: `John Doe changed the description`
**After**: `✏️5 ℹ️→❗`

### Quote Changes
**Before**: `John Doe changed the description` (displayed as `&gt; text`)
**After**: `✏️3 →❞ Important note`

### Combined Tags (Cleaner Output)
**Before**: `✏️5 <del>```</del><ins>-</ins><ins> </ins><ins>[x]</ins><ins> </ins><ins>ToDo</ins><ins> </ins><ins>1</ins>`
**After**: `✏️5 <del>```</del><ins>- [x] ToDo 1</ins>`

## 🧪 Testing

Extensively tested with various scenarios:
- ✅ Simple line moves and modifications  
- ✅ Complex multi-operation changes
- ✅ Checkbox state transitions
- ✅ Callout block type changes (info, success, warn, error)
- ✅ Quote line handling and `&gt;` fix
- ✅ Code block markers
- ✅ Empty content and edge cases
- ✅ Large content blocks and performance
- ✅ HTML tag combination and whitespace merging
- ✅ Activity feed compatibility (no more escaped HTML)

## 🚀 Benefits

### For Users
- **Clear Understanding**: See exactly what changed in card descriptions
- **Reduced Cognitive Load**: Filter out noise, focus on meaningful changes
- **Better Context**: Line numbers and move tracking provide better orientation
- **Professional Presentation**: Clean, modern emoji-based indicators
- **Proper Quote Display**: No more `&gt;` characters in activity feed
- **Cleaner Diff Output**: Combined HTML tags reduce visual clutter

### For Developers  
- **Maintainable Code**: Clean service architecture with separation of concerns
- **Dependency Injection**: Proper DI container registration for DiffService
- **Extensible Design**: Easy to extend for other content types
- **Performance Optimized**: Efficient algorithms with minimal overhead
- **Backward Compatible**: Existing functionality remains unchanged
- **Activity Feed Safe**: Handles all HTML escaping edge cases

## 🔧 Configuration

No configuration required - the feature works out of the box with existing Deck installations.

## 📋 Changelog

### Latest Updates
- **Quote Line Handling**: Fixed `&gt;` display issues in activity feed for quote lines
- **Combined HTML Tags**: Reduced visual clutter by merging consecutive `<ins>`/`<del>` tags
- **Whitespace Merging**: Improved tag combination to include whitespace between add/remove operations
- **Updated Emojis**: 
  - Error callout: 🔴 → ❗ (better matches UI)
  - Quote blocks: 💬 → ❞ (quotation mark)
  - Code blocks: 📝 → ‹› (guillemets)
- **Dependency Injection**: Proper DI registration for DiffService

### Code Review Feedback Addressed
- [x] Copyright year updated to 2025
- [x] Checkbox pattern extracted as class constant
- [x] Unused parameters removed
- [x] DiffService properly registered in DI container
- [x] Consistent line numbering (shows actual line numbers from content)
- [x] Activity feed compatibility verified and fixed

## 🤝 Contributing

This implementation focuses on card description changes but the `DiffService` architecture is designed to be extensible for other content types in the future.

---

*This PR represents a significant enhancement to user experience by providing clear, actionable insights into content changes while maintaining the professional quality expected from Nextcloud applications.*

putt1ck commented:
Looks interesting. Does it also address https://github.com/nextcloud/deck/issues/7058?

LexioJ commented:
> Looks interesting. Does it also address #7058?

Hard to tell as issue #7058 seems to discuss the overall appearance within the activity app and a wrong description notification when only comments were made. 

This PR indeed addresses the need to reduce bloating the activity log, but only within the activity tab of the corresponding card. It focuses on the deck app itself.

luka-nextcloud reviewed (changes requested):
Thank you for your contribution. The line number seems incorrect.

<img width="2078" height="1213" alt="image" src="https://github.com/user-attachments/assets/c99e6a2f-2c1d-4503-8beb-9a9f4acd4c0f" />

<img width="2197" height="1271" alt="image" src="https://github.com/user-attachments/assets/31c908b1-d664-410b-80ae-66adc8fa66b8" />

LexioJ commented:
> @LexioJ Let me know if this PR is ready for review.

Actually it is, I addressed all your remarks, improved the code & style further and being excited if this proposal will find its way to Deck for imho everyone’s benefit

LexioJ commented on lib/Service/DiffService.php:
```diff
@@ -0,0 +1,684 @@
+			} else {
+				// No good match, keep as remove
+				$enhancedOps[] = $removeOp;
+				$usedRemoves[] = $removeIndex;
```
I am sorry but I am not able to add you recommendations as I no longer understand the branches in this repo - seems I messed up my fork and local repo. If you like to use/integrate this feature I ask the nextcloud team to adapt/use the code I contributed

LexioJ commented on lib/Service/DiffService.php:
```diff
@@ -0,0 +1,684 @@
+					break;
+				case 'keep':
+					$word = $oldWords[$operation['old_line']] ?? '';
+					break;
```
I am sorry but I am not able to add you recommendations as I no longer understand the branches in this repo - seems I messed up my fork and local repo. If you like to use/integrate this feature I ask the nextcloud team to adapt/use the code I contributed

marcoambrosini commented:
Sounds like a good compromise to me! Or all other places could actually just link to the deck activity tab without detailing the exact changes. 
What do you think @LexioJ ?

LexioJ commented:
> Sounds like a good compromise to me! Or all other places could actually just link to the deck activity tab without detailing the exact changes. What do you think @LexioJ ?

I created a screenshot with annotations to illustrate what this code aims to achieve - if there are better ideas or suggestions, would be happy to read them 😀 - overall goal is to make it easier to understand what actually changed on a given modification within the description field.
<img width="995" height="943" alt="2025-12-15_22-12-41" src="https://github.com/user-attachments/assets/21f9e549-80c7-44d4-b367-1ff4ba5cd78d" />

- Intelligent Move Detection (🚚): Identifies when lines are relocated without modification
- Word-Level Modifications (✏️): Highlights specific word changes within lines using combined `<del>`/`<ins>`HTML tags for cleaner output (<ins>something added</ins> or <del>something deleted</del>)
- Clean Additions/Deletions (✨/🗑️): Clear indicators for new or removed content
- Checkbox State Tracking: Special handling for task list changes with visual emoji transitions ("☑️→🔲" or "🔲→☑️")
- Quote Line Handling: Properly formats blockquotes with ❞ emoji (fixes &gt; display issues)
- Callout Block Transitions: Shows emoji transitions for callout type changes (ℹ️→❗, ✅→⚠️, etc.)
- Code Block Indicators: Displays code block markers with ‹› symbol

marcoambrosini commented:
A few comments from my side
* I also think we should not reference lines for these changes. Line numbers are not a reference point for normal users
* Relying on emojis to show the type of change makes it a bit confusing. Especially if there are emojis in the text. I would avoid it
* I would limit the entries to addition (highlighted), deletions (strikethroughs) and checked, and . 
For checked I'd rather use a sentence instead: 
   - Alice marked `☑️ Task` as checked
   
@juliusknorr do you have some thoughts here?

marcoambrosini commented:
> HTML is not allowed in descriptions and messages of activity

To clarify here @nickvergessen, does this limitation apply for the activity app only, or also for the activity tab of the deck card?
